### ISSUE TITLE
Consider using ASCII only strings

### ISSUE THREAD

**Issue by mgostIH** [stadust/dash-rs#2]
# Reason 
- The game currently handles everything in ASCII only, supporting unicode would never work when interfacing with the game.
- We currently use **unsafe** only for `Vec` -> `String` conversion, it would be nice to remove this entirely, making the crate `#![forbid(unsafe)]`
- By using a `String` representation, we are forcing wasteful UTF-8 checking for the users and possibly for future internal parts of the library when iterating on a string.
For example, if we plan to support parsing parallelization, we'll need to find delimiters quickly in the middle of serialized data, which is strictly not possible if the representation is UTF-8.

# Reasons against this
This might require another dependency like [ascii](https://crates.io/crates/ascii) and to expose this type to the user. We technically only need the ASCII representation only internally, but as I said earlier giving back a String would force the user a stronger contract than needed.
**What crate to use exactly is up to decision.**

# Alternatives
We could simply use unsafe internally, but this makes the library less likeable and makes future code harder to verify. 
Moreover we'd need to be careful when **handling input data**, it's possible to explore a model where we are being attacked by carefully crafted input if, for example, this library happens to be on websites.

**Comment by stadust** [comment]
Alright so I've thought about this for a bit and arrived at the following conclusions

+ **For now**, our use of `std::string::from_utf8_unchecked` is safe because deserialization happens by parsing an `&str`, which is ensured to be valid UTF8 already. It is also currently not possible to insert invalid utf8 into one of our structs (modulo further use of unsafe of course)
+ If we ever switch to basing our Deserializer on a `Vec<u8>` (which might make sense considering the data we're processing is read from a socket or file) or similar, all bets are off and the above becomes unsound. (This is probably what you meant by being careful with input data). 
+ `#![forbid(unsafe)]` is generally a very desirable thing to have

However, I don't think we can say that the game only supports ASCII. The only thing indicating this is that the font files only consist of ASCII-chars. However, there is no server-sided validation that prevents unicode usage across most endpoints, and unicode is correctly handled by the game (as can be seen by gdbrowser.com, which posts comments with special unicode emoji to mark comments made using it. They simply don't show up in-game, but are handled correctly by the API). 

From my (admittedly limited) understanding of UTF-8, efficient parallelization is still possible efficiently because we can just jump to a random byte in the string and call `str::split_at` the next non-continuation byte, which is at most 3 bytes away (so this is a constant time operation). This thus wouldn't even require unsafe code.

In conclusion, switching to ASCII-only is (in my understanding) not possible, because we are not guaranteed to receive ASCII-only data and we should generally not be able to process less input than the official game client. Even if we decided to do that, some sort of validation that the input is ASCII-only would have to occur somewhere anyway (and since the game servers don't do it, we or our users would have to). Whether we do ASCII-validation or UTF8-validation doesn't make much difference, it'd both be O(n), so we might as well just use unicode.

**Comment by mgostIH** [comment]
Another option could be handing the user a Vec<u8> instead, allowing them to actually deal with the data, althought I'd expect this to be much simpler to use when sending it over a socket or saving to a file. This also doesn't require any unsafe or external crates, maintaining a lot of space for us regarding what we can optimize in the future.

**Comment by stadust** [comment]
as you mentioned on discord, serializing directly to a Writer is probably the most dynamic approach

**Comment by mgostIH** [comment]
Using a writer solves this problem for serialization and is generally a preferrable alternative to storing all the data in memory, so I agree this is solved on that end.
Deserialization however would still require all the data in memory, we can't stream deserialize because of lazy thunks and no allocation guarantees, but this allows for parallelization in splitting the data, which **can be done** even in UTF8, since as we discussed on Discord, searching for a single ASCII character (the separator) is well defined even across multiple unicode code points, so I am closing the issue, no specific ASCII needed.
